sysl2/sysl/seqs.go

```diff
@@ -0,0 +1,1081 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/anz-bank/sysl/src/proto"
+	"github.com/sirupsen/logrus"
+)
+
+type indentWriter struct {
+	bytes.Buffer
+	ind      int
+	complete bool // if the current position is a complete line
+}
+
+func (w *indentWriter) Write(p []byte) (n int, err error) {
+	newline := []byte("\n")
+	newlines := bytes.Count(p, newline)
+	if newlines == 0 {
+		if w.complete {
+			w.writeIndent()
+		}
+		n, err = w.Buffer.Write(p)
+		w.complete = false
+		return n, err
+	}
+
+	frags := bytes.SplitN(p, newline, newlines+1)
+
+	for i, frag := range frags {
+		if w.complete {
+			w.writeIndent()
+		}
+		nn, err := w.Buffer.Write(frag)
+		n += nn
+		if err != nil {
+			return n, err
+		}
+		if i+1 < len(frags) {
+			if _, err := w.Buffer.Write(newline); err != nil {
+				return n, err
+			}
+			n++
+		}
+	}
+	w.complete = len(frags[len(frags)-1]) == 0
+	return n, nil
+}
+
+func (w *indentWriter) WriteString(s string) (n int, err error) {
+	if !strings.Contains(s, "\n") {
+		if w.complete {
+			w.writeIndent()
+		}
+		w.complete = false
+		return w.Buffer.WriteString(s)
+	}
+	return w.Write([]byte(s))
+}
+
+func (w *indentWriter) writeIndent() error {
+	if !w.complete {
+		return nil
+	}
+
+	for i := 0; i < w.ind; i++ {
+		if err := w.Buffer.WriteByte(' '); err != nil {
+			return err
+		}
+	}
+
+	w.complete = false
+
+	return nil
+}
+
+func (w *indentWriter) indent() { w.ind++ }
```

Comment:
Make this a multiline function.

This is warranted if there's a family of single-statement functions, e.g.:
```go
func Add(a, b Something) Something { return Something{a.value + b.value} }
func Sub(a, b Something) Something { return Something{a.value - b.value} }
func Mul(a, b Something) Something { return Something{a.value * b.value} }
func Div(a, b Something) Something { return Something{a.value / b.value} }
```

In this case, it's not only on its own (and therefore won't save significant amounts of vertical space) but also now looks structurally different to its counterpart, `unindent()`, though they are two sides of the same coin.